scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/BucketMetadata.java

```diff
@@ -290,14 +290,47 @@ <K> byte[] encodeKeyBytes(K key, Coder<K> coder) {
   }
 
   // Checks for complete equality between BucketMetadatas originating from the same BucketedInput
-  public abstract boolean isPartitionCompatibleForPrimaryKey(BucketMetadata other);
+  public boolean isPartitionCompatibleForPrimaryKey(BucketMetadata other) {
+    return isIntraPartitionCompatibleWith(other, false);
+  }
+
+  public boolean isPartitionCompatibleForPrimaryAndSecondaryKey(BucketMetadata other) {
+    return isIntraPartitionCompatibleWith(other, true);
+  }
 
-  public abstract boolean isPartitionCompatibleForPrimaryAndSecondaryKey(BucketMetadata other);
+  private <MetadataT extends BucketMetadata> boolean isIntraPartitionCompatibleWith(
+      MetadataT other, boolean checkSecondaryKeys) {
+    if (other == null) {
+      return false;
+    }
+    final Class<? extends BucketMetadata> otherClass = other.getClass();
+    final Set<Class<? extends BucketMetadata>> compatibleTypes = compatibleMetadataTypes();
+
+    if (compatibleTypes.isEmpty() && other.getClass() != this.getClass()) {
+      return false;
+    } else if (this.getKeyClass() != other.getKeyClass()
+        && !(compatibleTypes.contains(otherClass)
+            && (other.compatibleMetadataTypes().contains(this.getClass())))) {
+      return false;
+    }
+
+    return (this.hashPrimaryKeyMetadata() == other.hashPrimaryKeyMetadata()
+        && (!checkSecondaryKeys
+            || this.hashSecondaryKeyMetadata() == other.hashSecondaryKeyMetadata()));
+  }
+
+  public Set<Class<? extends BucketMetadata>> compatibleMetadataTypes() {
+    return new HashSet<>();
+  }
 
   public abstract K1 extractKeyPrimary(V value);
 
   public abstract K2 extractKeySecondary(V value);
 
+  abstract int hashPrimaryKeyMetadata();
+
+  abstract int hashSecondaryKeyMetadata();
```

Comment:
the secondary key impl. (lexicographic) we have today is one way of implementing it, which in some cases like analytics is less efficient as opposed to sth like z-ordering. Might be nice to think a bit more to find out with a more future proof API e.g. support N keys?